### PR TITLE
Merge pull request #9229 from storybookjs/core/preview-layouts

### PR DESCRIPTION
**Pull request by ndelangen**
This adds support for a layout parameter: https://github.com/storybookjs/storybook/pull/8358#issuecomment-568295563

The parameter has 3 possible values: 'padded' (default), 'fullscreen' & 'center'

Here's how you use it:

```js
export default {
	title: 'my Component',
	parameters: {
		layout: 'fullscreen'
	}
}
```

This allows users to specify whether they want padding or now, and also deprecates addon-centered, making it a core-feature

### CODE REVIEW

**Review by ghengeveld — CHANGES_REQUESTED** [review]
I like the idea of a `layout` option, but I'm not so keen on this implementation. I don't think we need memoization and I would really like to have the possibility to provide a callback instead of a value so we can determine the value based on the story context.

I would really like to champion this change as soon as I'm back at work.

**Comment by ghengeveld** [comment]
I updated the branch, refactored some logic to avoid introducing a dependency and added support for layout based on story context.

**Comment by ghengeveld** [comment]
@ndelangen are you okay with the changes I made?

File: lib/core/src/client/preview/start.js [hunk 1]
```diff
@@ -122,24 +121,31 @@ function focusInInput(event) {
...
+      margin: 0;
+      padding: 1rem;
+    `,
+  }[layout];
+  if (styles && layout !== previousLayout) {
+    document.body.style = styles;
+    previousLayout = layout;
   }
-});
+};
```
Comment (ndelangen) [inline lib/core/src/client/preview/start.js]: Why is this better over a memoized function?

File: examples/official-storybook/stories/core/layout.stories.js [hunk 2]
```diff
@@ -1,43 +1,35 @@
...
+// eslint-disable-next-line react/prop-types
+const Box = ({ children, display = 'block' }) => (
+  <div style={{ display, border: '2px solid #FF4785', padding: 10 }}>{children}</div>
+);
 
 export default {
   title: 'Core/Layout',
   parameters: {
-    layout: 'padded',
+    layout: ctx => ctx.name.toLowerCase().split(' ')[1],
```
Comment (ndelangen) [inline examples/official-storybook/stories/core/layout.stories.js]: I'd rather not do this, until it's really required/requested from the community.

I think this is too "magical" tbh.

Comment (ghengeveld) [reply examples/official-storybook/stories/core/layout.stories.js]: Yeah okay, maybe not in official-storybook. I did this to demonstrate how the callback works. I guess a (unit) test is a better place for it.

File: lib/core/src/client/preview/start.js (hunk 1 above)
Comment (ghengeveld) [reply lib/core/src/client/preview/start.js]: No external dependency.

Comment (ghengeveld) [reply lib/core/src/client/preview/start.js]: Also memoization is meant to cache the result of a function, not prevent a side-effect from running multiple times. Memoization is not the right tool for that.

File: examples/official-storybook/stories/core/layout.stories.js (hunk 2 above)
Comment (ndelangen) [reply examples/official-storybook/stories/core/layout.stories.js]: I just don't foresee people needing the callback, and it's another bit of complexity.

Functions are harder to share between manager & preview. So I'd rather not add this function option if it's not required by the community.

**Review by ghengeveld — APPROVED** [review]

**Comment by ghengeveld** [comment]
🎉

**Comment by shilman** [comment]
🎉🎉🎉

@ndelangen @ghengeveld is there a corresponding documentation and/or MIGRATION update that's missing? 🙏

**Comment by ndelangen** [comment]
@shilman YES, I think we should add some proper documentation for this.

**Comment by frebro** [comment]
Trying to get this working for MDX with Vue. Is it supposed to look like this? 

```
<Story name="MyStory" layout="fullscreen">
  {{
    components: { MyComponent },
    template: `
      <MyComponent />
    `
  }}
</Story>
```

**Comment by frebro** [comment]
Trying to get this working on MDX with Vue. Is it supposed to look like this? 

```
<Preview layout="fullscreen">
  <Story name="MyStory">
    {{
      components: { MyComponent },
      template: '<MyComponent />'
    }}
  </Story>
</Preview>
```

I'm using "next" (currently 6.0.0-alpha.18) for everything @storybook.

**Comment by ndelangen** [comment]
I'll investigate this @frebro

**Comment by frebro** [comment]
Tried with your example in the PR above @ndelangen, and it works fine. Hartelijk bedankt!

**Pull request merged**